### DISCUSSION TITLE
Maccy 1.x and towards 2.x

### DISCUSSION THREAD

**Discussion by p0deje** [p0deje/Maccy#790] [Announcements]
> **TLDR**
> I plan to freeze current Maccy to 1.x version and start re-writing it to utilize newer technologies, which will unfortunately necessitate dropping support for older macOS versions. The new version will maintain the same look and feel as the current version, while being faster, more efficient and have better integration with modern Apple ecosystem.

> **UPDATE 1** (July 8, 2024)
> Maccy 1.0 has been released - see https://github.com/p0deje/Maccy/discussions/790#discussioncomment-9991320.

> **UPDATE 2** (September 16, 2024)
> Maccy 2.0 has been released - see https://github.com/p0deje/Maccy/releases/tag/2.0.0

Hello folks! 👋 

Maccy was created 7 years ago as a hobby project where I attempted to learn Swift and macOS development to replace missing Parcellite clipboard manager when I moved from Linux to macOS. Since then, it has been download 600k+ times, got featured on the App Store and used by thousands of people every day. I've never thought it might become so popular. Thank you everyone for using it, it's an honor for me. 🙇 

7 years is a lot of time and macOS has changed drastically since then. When I started, there were no SwiftUI, SwiftData and other modern approaches to developing AppKit applications. The way Maccy is built is based on the old NSMenu system that is used to power context menus in macOS. This allowed me to preserver plenty of backwards compatibility - Maccy still runs and works on older macOS 10.14 Mojave which was released in 2018. There are still some people downloading and using it on those versions, though it's just a fraction of usage compared to modern macOS. 

The choice of NSMenu provided native UI so that Maccy is fast and feels like part of OS. However, it also caused multiple issues in the long run, some of which were worked around (e.g. support for Korean/Chinese language in search), while others are proven to be hard to work around (e.g. support for Japanese language). In summary, NSMenu were never designed to be used for such use cases and this caused plenty of problems with no straightforward way to work them around. I can elaborate more on these issues if necessary.

With all this technical debt and design limitations, I don't see a clear way forward rather than attempting to rewrite Maccy into a regular macOS application that uses regular windows and custom UI views for handling input events. In the very end, UI must remain the same so it's native and fast, but the underlying implementation must move away from NSMenu.

However, re-writing without using modern stack is quite painful so I'd like for Maccy to adopt all the recent SDKs by Apple, notably:
- SwiftUI for UI views instead of current AppKit;
- SwiftData for persistence instead of current Core Data;
- App Intents for Shortcuts.app integration instead of current Intents.

However, switching to these technologies comes at the price of dropping backwards compatibility. For example, SwiftData requires macOS 14 Sonoma or later. That's unavoidable even though I'd hate to do that. The installations per platform suggest that's acceptable - by the time rewrite is done the 90% of users would be using at least macOS 14.  I don't have data for installations via Homebrew/GItHub, but on the App Store the downloads this year so far are the following:

| macOS version | Downloads (%) |
| ------- | --------------- |
| 10.14 | <1% |
| 10.15 | <1% |
| 11 | <1% |
| 12 | 4% |
| 13 | 9% |
| 14 | 87% |

With that said, my current plan is to:
1. Make a final release of current Maccy as version 1.0. Keep it supported for a while with bug fixes whenever it's possible.
2. Start working on Maccy 2.0 that uses SwiftUI, SwiftData, App Intents. Preserve current UI look and feel while solving number of issues (memory consumption, Japanese language support). 
3. Explore opportunities that open with new technologies (iCloud synchronization, iOS version, integration with new Siri, etc).

I'd be happy to hear what Maccy users think about these plans!

**Comment by XInTheDark** [comment]
Sounds great! Looking forward to the next version!

**Reply by AJ-xylo** [reply]
I dont know why but i am unable to download the Maccy 2.0 version. I tried to restart and redownload it, nothing works. Could someone help me with it?

![Uploading image.png…]()

**Reply by Justbeginwithit** [reply]
same

**Reply by Alrazzouq** [reply]
Me too.

**Reply by jtphoenixga** [reply]
I'm also a part of the consesnsu, not able to download Maccy 2.0

**Reply by p0deje** [reply]
Sorry, that's because I pushed the update without releasing it. Please try again tomorrow!

**Comment by tommyming** [comment]
Sound great! As a "maccy-heavy" user, would say it helps me a lot on software development.
But one thing I am not sure about is SwiftData. 

Although it works well with new tech stack e.g. SwiftUI, the issues on new `Predicate` system could be a concern.
Apart from that, I've heard that SwiftData has a huge change in this year's WWDC(I don't have the time to drill on the related sessions yet, sadly). 
Not sure how much will it change the behavior of SwiftData 1.0.

**Reply by p0deje** [reply]
> Although it works well with new tech stack e.g. SwiftUI, the issues on new Predicate system could be a concern.

What issues are you referring to? Maccy is not using any complicated data queries so I believe it should be okay.

**Reply by tommyming** [reply]
Thanks for the reply!

**Disclaimer**: I didn't dive deep into the code, so my comment is just opinon.

I faced some issues when getting my hands dirty on SwiftData when it was first released last year.
It is slightly different to `NSPredicate`. 

My assumption right now is: 
If Maccy is just performing a simple query when checking copy history, then it should be fine.

I can't give a very subtle opinion at this moment. But I agree that it is worth trying.

**Comment by edsonbittencourt** [comment]
It was love at first sight with Maccy. I would love to and am really excited about the intentions for the application. Congratulations.

**Comment by miodragnikac** [comment]
Looking forward to the iOS version 🙌

**Comment by weisJ** [comment]
That’s great news. Let me know if you need any help. I still have cursor positioning in my backlog :D

**Comment by CostGranda** [comment]
That's amazing news, go ahead! and thanks for the effort

**Comment by p0deje** [comment]
Alright, folks, I've had some progress with 2.x and I am now convinced it's possible to migrate to App Intents + SwiftData + SwiftUI for Maccy. 

In anticipation of 2.x, here is the promised 1.0.0 version - https://github.com/p0deje/Maccy/releases/tag/1.0.0. It's pretty much the same old Maccy with one major storage change that **prevents older versions of Maccy from working with the storage after the upgrade**. I'm fairly certain that's the only breaking change I need to do before 2.x, so 1.x upgrades should be smooth further on, respecting the semantic versioning.

**Reply by qupig** [reply]
If I understand correctly, the point being made here is that once you have upgraded to version 1.0 and are running once, then if you downgrade to version 0.x again it will not work.

If that's the case, the wording in the release may need to be more explicitly stated, as that looks more like a warning that upgrading from 0.x to 1.0 will not work properly and will require some manual handling.

People are unlikely to use both version 0.x and 1.0 at the same time (hence why we call it an upgrade) and most won't be downgrading to 0.x again, so the warning's prominent placement and somewhat ambiguous wording can make it easy to misinterpret.

Finally, thank you for making this open source software.

**Reply by p0deje** [reply]
> This is the first release in 1.x version branch. Please note, that older versions of Maccy (0.x) will not work correctly after the upgrade.

Is this confusing? Some people might want to revert to 0.x if there is a bug in 1.0.

**Reply by twknab** [reply]
This language makes sense to me and isn't confusing. Thanks for all your hard work on this, super excited and love Maccy!

**Comment by taylor325** [comment]
Sounds great!

**Comment by tysecure** [comment]
Any privacy concerns we should be aware of?

**Reply by p0deje** [reply]
What kind of concerns? 

There would be no changes in data collection - Maccy is not collecting any data about users, all the copied items are left on your computer. Maccy is running in App Sandbox which restricts network access, so it cannot send anything by design. The only exception for communication with external networks is downloading application updates.

Adopting SwiftData would likely allow using iCloud for storage and synchronization of data between multiple Macs (#182). This would still use Apple's CloudKit and be opt-in.

**Reply by tysecure** [reply]
Ok. Thanks for clarifying.

**Comment by akoblentz** [comment]
Thank you!

**Comment by natedame** [comment]
Sounds great! Huge fan. I'm not a developer, but if I can help some other way, let me know.

**Comment by arunim2405** [comment]
I would like to contribute, big fan of the project, but I will need some guidance in the start 😃

**Reply by p0deje** [reply]
Feel free to grab any interesting issue that is marked with a "help wanted" label and ask specific questions, I'll do my best to guide you.

**Comment by p0deje** [comment]
If anyone is up to try a new re-written beta version, please take a look at https://github.com/p0deje/Maccy/discussions/818. This version can be run side-by-side with Maccy 1.0, so you should be able to test easily. It would be great if you could share the feedback on:

- How the new UI feels like.
- Performance issues.
- Memory utilization concerns.
- Dvorak or non-US keyboard layout issues.
- Any other issues not listed in **Known issues**.

**Comment by zrf2v0m** [comment]
This app is very good. It is a tool I have used for a long time. Does the 2.0 version consider cross-platform support, such as linux and windows?

**Reply by p0deje** [reply]
No, that would require rewriting Maccy to use a cross-platform UI library such as QT. That's not my plans. However, 2.0 would open a possibility of porting Maccy to other operating systems in Apple ecosystem (iOS, IPadOS, etc)

**Reply by luckydonald** [reply]
I'm not sure iOS (-Appstore) would officially allow some app like that.
There's been **Clip** in the Altstore for a while, but that running in the background stuff is really not something iOS likes.

Article about Clip: https://www.macstories.net/reviews/altstores-clip-is-the-best-clipboard-manager-on-ios-yet/

**Comment by bgpntx** [comment]
Hey, thanks for one of the most useful app, happy user for ~5 yrs. Just wanted to let you know 2.0 is failing to start on fresh 15.0 Beta (24A5298h). Keep on working and thanks again!

**Reply by p0deje** [reply]
Hmm, can you share the logs fron Console.app?

**Reply by bgpntx** [reply]
I get just segmentation fault(
<img width="858" alt="Screenshot at Jul 24 08-21-35" src="https://github.com/user-attachments/assets/ec1b89ce-c0fd-4f4b-bcab-bb2e2883213d">

**Reply by bgpntx** [reply]
[Maccy2-0-err.txt](https://github.com/user-attachments/files/16357948/Maccy2-0-err.txt)
Here's log from Console.app

**Reply by p0deje** [reply]
Please try beta 2 - https://github.com/p0deje/Maccy/releases/tag/2.0.0.beta.2 - it should be fixed there.

**Reply by bgpntx** [reply]
Yup, fixed. Thanks a lot!

**Comment by Lui22** [comment]
For some reason the window is too narrow. And the 1.0.0 version too. The previous version I was using is 0.29.something. I would like to share any info to research this

<details><summary>Screenshot</summary>![image](https://github.com/user-attachments/assets/737bc5c4-fef2-4df4-b0e7-bcf0b501d830)

<p>

![image](https://github.com/user-attachments/assets/737bc5c4-fef2-4df4-b0e7-bcf0b501d830)


</p>
</details>

**Reply by p0deje** [reply]
@Lui22 Check "Title length" setting in Advanced.

**Reply by Lui22** [reply]
@p0deje it is set to 50

**Reply by p0deje** [reply]
@Lui22 Can you try changing it to 51 when running Maccy 1.0?

**Reply by Lui22** [reply]
@p0deje changed and it fixed :)

**Reply by Lui22** [reply]
@p0deje oh but the pinned one is still collapsed

**Reply by p0deje** [reply]
@Lui22 You'd need to edit its title via Pins settings, not sure what caused this in the first place.

**Reply by Lui22** [reply]
But it looks like a migration problem only. I unpinned and pinned it again and now everything is ok

**Comment by p0deje** [comment]
It’s not released yet.

**Comment by nsyris2** [comment]
Version 2 of Maccy is saying it's available to update to via Maccy in-app update but it is not working...

**Reply by p0deje** [reply]
Sorry, that's because I pushed the update without releasing it. Please try again tomorrow!

**Comment by NikoMuffin** [comment]
Paging @p0deje   Update mechanism in Maccy is broken FYI

@nsyris2

**Reply by p0deje** [reply]
Sorry, that's because I pushed the update without releasing it. Please try again tomorrow!

**Reply by NikoMuffin** [reply]
> Sorry, that's because I pushed the update without releasing it. Please try again tomorrow!

Absolutely no worries. I wanted to say thank you from the bottom of my heart for maintaining and having Maccy available to all for free. When I get a job I put you on my donate list. ❤️

**Comment by p0deje** [comment]
2.0.0 is finally out, you can upgrade. Please share your feedback and open issues for bugs you encounter or missing features.

**Comment by TickDracy** [comment]
While I think Maccy is a great app, I'm not a fan of using non-stable OS's (yea, I'm looking at you Sonoma), so I don't plan to move to that System for the next two years (yes, I'm really that drastic).  

My question is: are there any plans to update version 1.x with bug fixes, or it's the _final_ version?

**Reply by p0deje** [reply]
@TickDracy Do you have specific bugs you would like to be fixed? Many of them are limitations of 1.0 architecture and cannot be fixed without a rewrite such as 2.0.

**Reply by TickDracy** [reply]
> @TickDracy Do you have specific bugs you would like to be fixed?

Hi @p0deje, no, I don't have any specific bug. It was just a question to analyze the future of the app, and if I should continue using it, or find an alternative, given that I've recently moved into macOS Ventura, and I will (definitely) take a long time until I reach the version supported by Maccy 2.0.  

My only annoyance (not with Maccy, more of a Bitwarden issue) is that when I copy passwords from the browser extension, they get added into Maccy. While I can avoid this on the Desktop application (by adding it to the exclusion), I don't find a way to do this for the browser extensions. But like I said, I don't think this is an issue with Maccy itself...

**Reply by luckydonald** [reply]
Yeah, that last part is super annoying, but probably it can only be solved on Bitwarden side, if at all.

**Reply by p0deje** [reply]
> Hi @p0deje, no, I don't have any specific bug. It was just a question to analyze the future of the app, and if I should continue using it, or find an alternative, given that I've recently moved into macOS Ventura, and I will (definitely) take a long time until I reach the version supported by Maccy 2.0.

I would consider 1.0 a fairly stable release with the only issues being related to architecture (cannot popup over password fields, memory utilization is not optimal, application focus issues). Those are not going to be fixed, but I can try fixing other issues in 1.0 unless they require a lot of work.

> My only annoyance (not with Maccy, more of a Bitwarden issue) is that when I copy passwords from the browser extension, they get added into Maccy. While I can avoid this on the Desktop application (by adding it to the exclusion), I don't find a way to do this for the browser extensions. But like I said, I don't think this is an issue with Maccy itself...

That's a common issue for password managers used from the browser - there is no simple way to distinguish them from other copies coming from the browser itself.

**Reply by TickDracy** [reply]
Hey @p0deje, I guess Apple's lack of quality on macOS took me by surprise. I should have waited for someone to test the latest Ventura update (I was not imaging they were breaking things on such an old system), but it [messed up my system](https://forums.macrumors.com/threads/brightness-controls-no-longer-working-after-13-7-update.2436583/?post=33406621#post-33406621), so my only solution (aside from starting from scratch and trying to restore using Time Machine) was to update to macOS 14.  

I guess I have to eat my own words on the original plan of _using Ventura for more than one year_. The positive side is that I've got to upgrade Maccy to version 2.0. Not everything is negative in this story, I guess...

**Comment by alexandermyasnikov** [comment]
It's amazing how fast the search works.
In version 1.0 mixed search occasionally slowed down on 999 records. But in 2.0 it works instantly ;-)

**Comment by codecantata** [comment]
@p0deje 
I noticed that **Antonia Ivanova** seems to be linked to the distribution of the Maccy app on the Mac App Store. However, I was unable to find any details regarding why the app is not distributed by yourself. Could you kindly provide some clarification on this?

**Reply by p0deje** [reply]
This is the correct distribution team - I could not do it myself back then due to personal reasons.

**Reply by codecantata** [reply]
Alright, thank you for confirming 👍🏻

**Comment by GianMarco-Tavazzani** [comment]
Dear Alex, first THAN YOU! I'm still using Maccy 0.16 with satisfaction on High Sierra; is there a following version for Mojave?
Again thanks by GianMarco Tavazzani

**Reply by p0deje** [reply]
Version 1.0 requires Mojave, while 2.0 already needs Sonoma.

**Comment by GianMarco-Tavazzani** [comment]
Dear Alex.

		you are very kind as ever 🤗, not only very clever as I appreciate by years!

		Thank you so much for your quick and gentle response!

    				GianMarco Tavazzani

				via Bergamo, 10 - I-20900 Monza (MB)
		whatsapp, Telegram (much better!), phone: +39 329 536 55 76
	Skype: gianmarcotav, Google Chat: ***@***.*** com

**Comment by whyzheng** [comment]
Hi Alex,

Does Maccy not work on macOS 15? I set the shortcut (⌥ + V) in the "General" settings to open it, but it doesn't take effect.

**Reply by qupig** [reply]
Apple prohibits all store apps from continuing to use `option ⌥` as a shortcut key combination.
